---
title: "TNBC Pharmacogenomic analysis"
---
Load basic libraries that are used throughout the notebook
```{r message=FALSE, warning=FALSE, include=FALSE}
rm(list=ls())
require("plotly")
require("matrixStats")
require("ComplexHeatmap")
require("seriation")
require("dplyr")
require("pathfindR")
require("edgeR")
require("org.Hs.eg.db")
require("dplyr")
```

plotly annotation tools
```{r}
vline <- function(x = 0, color = "black") {
  list(
    type = "line",
    y0 = 0,
    y1 = 1,
    yref = "paper",
    x0 = x,
    x1 = x,
    line = list(color = color, dash="dot"))

}

hline <- function(y = 0, color = "black") {
  list(
    type = "line",
    x0 = 0,
    x1 = 1,
    xref = "paper",
    y0 = y,
    y1 = y,
    line = list(color = color, dash="dot"))
}

anno_gen = function(df){
  library(ComplexHeatmap)
hm_anno = HeatmapAnnotation(df = as.data.frame(df_anno),
                            annotation_name_side = "left",
                            col=list(
                              Time=c("PRE"="grey80",
                                     "MID"="grey40",
                                     "POST"="grey0"
                                          ),
                              PiD =c("TNBC117"="yellow3",
                                     "TNBC231"="wheat3",
                                     "TNBC249"="violetred3",
                                     "TNBC283"="turquoise3",
                                     "TNBC010"="tomato3",
                                     "TNBC139"="thistle3",
                                     "TNBC047"="tan3",
                                     "TNBC271"="steelblue3",
                                     "ATID099"="springgreen3",
                                     "ATID608"="snow3"
                                     )
                              )
                            )
  return(hm_anno)
}

pathfindR_visNetwork = function(pfr_tmp,flnm="visNetwork.html"){
  ID_column <- "Term_Description"

  edges <- data.frame()
  all_up = c()
  all_down = c()
  for (i in base::seq_len(nrow(pfr_tmp))) {
    up_genes <- unlist(strsplit(pfr_tmp$Up_regulated[i],", "))
    all_up = unique(c(all_up,up_genes))
    down_genes <- unlist(strsplit(pfr_tmp$Down_regulated[i],", "))
    all_down = unique(c(all_down,down_genes))
    genes <- c(up_genes, down_genes)
    for (gene in genes) {
      edges <- rbind(edges, data.frame(Term = pfr_tmp[i, ID_column], Gene = gene))
    }
  }
  
  pathways = as.data.frame(unique(edges$Term))
  colnames(pathways) = "id"
  pathways[,"color"] = "grey"
  pathways[,"group"] = "Pathway"
  pathways[,"shape"] = "square"
  pathways[,"font.size"] = 20
  
  all_up = as.data.frame(all_up)
  colnames(all_up) = "id"
  all_up[,"color"] = "green"
  all_up[,"group"] = "Up_regulated"
  all_up[,"shape"] = "dot"
  all_up[,"font.size"] = 10
  
  all_down = as.data.frame(all_down)
  colnames(all_down) = "id"
  all_down[,"color"] = "darkred"
  all_down[,"group"] = "Down_regulated"
  all_down[,"shape"] = "dot"
  all_down[,"font.size"] = 10
  
  nodes = rbind(pathways,all_up,all_down)
  
  colnames(edges) = c("from","to")
  
  rm("pathways","all_up","all_down")
  
  graph = visNetwork(nodes, edges,width='100%',height=2000)%>%
    visPhysics(stabilization = list(iterations=500))%>%
    visNodes(color = list(inherit=FALSE), font = list(size = 30))%>%
    visEdges(color = list(inherit=FALSE, font = list(size = 40)))%>%
    visOptions(highlightNearest = list(enabled = T, degree = 2, hover = T)) %>%
    visInteraction(multiselect = TRUE) %>%
    visLayout(randomSeed = 123)%>%
    visGroups(groupname = "Pathway", color = "grey") %>%
    visGroups(groupname = "Up_regulated", color = "green") %>%
    visGroups(groupname = "Down_regulated", color = "darkred") %>%
    visLegend(position = "right")
  
  visSave(graph, file=flnm, selfcontained = TRUE, background = "white")
}

hsa_preview = function(pfr_results, input_df_tmp ,outdir="./hsa_preview"){
  
  pathview_in = input_df_tmp %>% dplyr::select("logFC")
  
  for(i in 1:nrow(pfr_results)){
    tryCatch({
    mypathway<-pfr_results$ID[i]
    Pathway_name = pfr_results$Term_Description[i]
    dir.create(file.path(outdir, Pathway_name), showWarnings = FALSE,recursive = TRUE)
    setwd(file.path(outdir, Pathway_name))
    library(pathview)
    pathview(gene.data=pathview_in,
             species="hsa",
             pathway=mypathway,gene.idtype = "SYMBOL",
             low = list(gene = "red", cpd = "blue"),
             mid = list(gene = "gray", cpd = "gray"), 
             high = list(gene = "green", cpd = "yellow"))},error=function(e){})
  }
}
```


Set working directory and read in the experimental data frames, select the cohort (matched pairs)
```{r message=FALSE, warning=FALSE}
library("readxl")
library("dplyr")
#library("xlsx")
library("dplyr")

Base_dir = getwd() #"D:\\Users\\rpowell\\Documents\\GitHub\\TNBC-PGx"
knitr::opts_knit$set(root.dir = Base_dir)
setwd(Base_dir)


#DNA_all <- read.xlsx2("data/TNBC_oncoprint_binary.xlsx",sheetIndex = 1,row.names = 1)
RNA_all <- read.csv("./data/Primary/RNA_combat.220304.csv",row.names=1)
#ssGSEA_all <- read.csv("./data/RNA_combat.220304.csv",row.names=1)

Drug_all <- read.csv("./data/Primary/Drug_AUC.csv",row.names=1)
Drug_meta <- read.csv("./data/Primary/Drug_metadata.csv",row.names=1)

Meta_all <- read_excel("./data/Primary/PDX_metadata.xlsx")
row.names(Meta_all) = Meta_all$PDX_ID

PPI = "./data/db/STRING.csv"

pairsOnly = T
cellLine_drug = colnames(Drug_all)
drugnames = row.names(Drug_all)

cellLine_RNA = colnames(RNA_all)
#genenames = row.names(RNA_filtered_RWR)

all_samples = intersect(cellLine_drug,cellLine_RNA)

Pairs = Meta_all %>% dplyr::count(PiD)
Pairs = Pairs[which(Pairs$n >= 2),]
Meta_pairsOnly = Meta_all[which(Meta_all$PiD %in% Pairs$PiD),]
row.names(Meta_pairsOnly) = Meta_pairsOnly$PDX_ID

if (pairsOnly){
  train_samples = intersect(all_samples,Meta_pairsOnly$PDX_ID)
} else{train_samples = all_samples}
train_samples

tmp = setdiff(colnames(RNA_all),train_samples)
test_samples = intersect(colnames(Drug_all),tmp)
unknown_samples = setdiff(tmp,colnames(Drug_all))

with_held_RNA = RNA_all[,test_samples]
with_held_Drug = Drug_all[,test_samples]
with_held_common = intersect(colnames(with_held_RNA),colnames(with_held_Drug))

with_held_RNA = with_held_RNA[,with_held_common]
with_held_Drug = with_held_Drug[,with_held_common]
rm("tmp","Pairs")
```
This section generates basic statistics and QC visualizations for RNAseq data.
```{r message=FALSE, warning=FALSE}
library("matrixStats")
RNA_means = rowMeans(as.matrix(RNA_all))
RNA_stdDev = rowSds(as.matrix(RNA_all))
plot(RNA_means,RNA_stdDev)

RNA_filter = which(RNA_means >= .5)
RNA_filtered = RNA_all[RNA_filter,]
RNA_filtered_z = t(scale(t(RNA_filtered), center = T, scale = T))

library("RColorBrewer")

nsamples <- ncol(RNA_filtered)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(as.matrix(RNA_filtered[,1])), col=col[1], lwd=2, ylim=c(0,1), las=2, main="", xlab="")
title(main="Filtered data", xlab="RNAseq")
#abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(as.matrix(RNA_filtered[,i]))
  lines(den$x, den$y, col=col[i], lwd=2)
}

boxplot(RNA_filtered[,], las=2, col=col, main="")
title(main="Filtered data",ylab="RNAseq")
```

```{r}
library(dplyr)
set.seed(1)
Meta_all %>% 
  group_by(TNBCtype) %>%
  summarise(no_rows = length(PDX_ID))
```


```{r}
library(ComplexHeatmap)
library(tidyr)
library(dplyr)

#Build pivot.df from metadata
df = as.data.frame(Meta_all[which(Meta_all$Set != ''),] %>% pivot_wider(id_cols = PiD, 
                                                                        names_from = Time,
                                                                        values_from = TNBCtype4))
row.names(df) = df$PiD
df = df[,-1]
#df <- df %>% replace(is.na(.), 'UNS')

#TNBCtype2Color dictionary
colors = structure(c('springgreen','springgreen4','skyblue1','skyblue4','darkorchid3','firebrick','grey40'), 
                   names = c('BL1','BL2','M','MSL','LAR','IM','UNS'))

Heatmap(df,
        col=colors,
        border = T,
        rect_gp = gpar(col = "black"))
```

Next, we looked at differential drug susceptibilities identified by the high throughput chemical screening assays. 
```{r}
library(matrixStats)
drug_range = rowRanges(as.matrix(Drug_all),useNames = TRUE)
drug_range[drug_range < 0] <- 0
drug_range = drug_range[,2]-drug_range[,1]

drug_filter = which(drug_range >= 0.5)
```

From the differential gene expression analysis comparing matched pairs, we observed enrichment of cell cycle related genes. This observation is consistent with the growth analysis performed on the ex vivo cultures from the high throughput screens, which show increases in growth rates when comparing the post-NACT to the pre-NACT PDX models. It is also well established that alterations in growth rate and confound PGx associations. Accordingly, in this section we performed a correlative analysis of the growth data to the AUC activities.
```{r}
ord = colnames(Drug_all)
gr = Meta_all[ord,]
test = cor(gr$Log2_FC_3day,t(Drug_all))
r2 = test^2

#get list of correlated and variable drug names
growth_corr = names(r2[,which(r2 > 0.25)])
variable_drugs = names(drug_filter)
x = list(Growth_correlated_drugs = growth_corr,
         Variable_drug_response = variable_drugs)

library(ggvenn)
ggvenn(x,
       fill_color = c("#0073C2FF","#CD534CFF"),
       stroke_size = 0.5, set_name_size = 4)
rm('x','r2','test','gr')
```
This is the list of variable drugs that are potentially confounded by the rate of growth
```{r}
intersect(growth_corr,variable_drugs)
```
And this is the list of drugs that are not correlated with growth.
```{r}
drug_filter = variable_drugs[!(variable_drugs %in% growth_corr)]
drug_filter[order(drug_filter)]
```
Drugs that are different between subtype
```{r fig.height=7, fig.width=16}
library(circlize)

ord = Meta_all$PDX_ID
Drug_selected = Drug_all[drug_filter,ord]

TNBCtype = Meta_all$TNBCtype

#Drop UNS
# select = which(!str_detect(TNBCtype, "IM|UNS"))
# Drug_selected = Drug_selected[,select]
# TNBCtype = TNBCtype[select]

AOV_P = c()
for(i in 1:nrow(Drug_selected))
{
  df.test = as.data.frame(t(rbind(Drug_selected[i,],TNBCtype)))
  colnames(df.test) = c("Drug_val","Factor")
  df.test = transform(df.test,Drug_val=as.numeric(as.character(Drug_val)))
  result = summary(aov(Drug_val~Factor,df.test))
  AOV_P[i] = result[[1]][["Pr(>F)"]][1]
}

font_size = 15



df = Drug_selected[which(AOV_P<0.05),]
rAnno = rowAnnotation(df=Drug_meta[row.names(df),"Target"])
col_fun = colorRamp2(c(0,1.0), c("blue","orange"))

Heatmap(df,
        column_split = TNBCtype,
        column_title_gp=gpar(fontsize=font_size,fontface = "bold"),
        row_names_gp = gpar(fontsize = font_size),
        row_names_max_width = max_text_width(rownames(df),gp = gpar(fontsize = font_size)),
        left_annotation = rAnno,
        row_split = Drug_meta[row.names(df),"Class"],
        row_title_rot = 0,
        column_names_gp = gpar(fontsize = font_size),
        column_names_max_height = max_text_width(colnames(df),gp = gpar(fontsize = font_size)),
        name = "AUC",
        col = col_fun,
        heatmap_legend_param = list(title_gp = gpar(fontsize = font_size,fontface = "bold"), labels_gp = gpar(fontsize = font_size))
        )
```
```{r}
Drug_meta[row.names(df),]
```

```{r}
# Load the necessary library
library(ggplot2)

df= c()

df = as.data.frame(cbind(TNBCType=TNBCtype,AUC=t(Drug_selected["PEVONEDISTAT_Broad",])))
df$PEVONEDISTAT_Broad = as.numeric(df$PEVONEDISTAT_Broad)

# Assuming df is your dataframe and it has columns AUC and TNBCType
p <- ggplot(df, aes(x=TNBCtype, y=PEVONEDISTAT_Broad))

# Create the violin plot
p <- p + geom_violin()

# Add individual data points
p <- p + geom_jitter(width=0.2)

# Display the plot
print(p)
```


```{r fig.height=5, fig.width=16}
ord = Meta_all$PDX_ID
Drug_selected = Drug_all[drug_filter,]

AOV_P = c()
for(i in 1:nrow(Drug_selected))
{
  df.test = as.data.frame(t(rbind(Drug_selected[i,ord],Meta_all$NACT_Response)))
  colnames(df.test) = c("Drug_val","Factor")
  df.test = transform(df.test,Drug_val=as.numeric(as.character(Drug_val)))
  result = summary(aov(Drug_val~Factor,df.test))
  AOV_P[i] = result[[1]][["Pr(>F)"]][1]
}

df = Drug_selected[which(AOV_P<0.05),ord]
Heatmap(df,
        column_split = Meta_all$NACT_Response)
```


```{r}
library("dendextend")
library("ComplexHeatmap")

CCL_corr = cor(as.matrix(t(scale(t(Drug_all[drug_filter,train_samples]),center=TRUE, scale=TRUE))),method = c("pearson"))
df_anno = Meta_all[train_samples,]%>%dplyr::select("Time","PiD")

K = 7
hc = hclust(dist(CCL_corr))
clus = cutree(hc,k=K)

df_anno_clust = cbind(df_anno,clus)
chisq.test(df_anno_clust$PiD, df_anno_clust$clus, correct=FALSE)
table(df_anno_clust$PiD, df_anno_clust$clus)

row_dend = as.dendrogram(hc)
row_dend = color_branches(row_dend, k = K)

HM = Heatmap(CCL_corr,
             name = "Pearson",
             top_annotation = anno_gen(df_anno),
             show_row_names = FALSE)
HM
rm('HM','df_anno','hc','df_anno_clust','CCL_corr','den')
```

Differential drug susceptibility analysis
```{r warning=FALSE}
library(edgeR)
metadata = Meta_all[train_samples,]
metadata[] <- lapply(metadata, factor)

design = model.matrix(~0+Time+PiD,metadata)
Rx_fit = lmFit(Drug_all[drug_filter,train_samples],design)
Rx_fit_eBayesfit = eBayes(Rx_fit,trend = TRUE,robust = TRUE)
Rx_top = topTable(Rx_fit_eBayesfit,coef = "TimeMID" ,adjust.method="fdr", number=Inf)

Mid_v_Pre = c()
Pst_v_Pre = c()
for (i in 1:nrow(Drug_all)){
  tmp1 = t(Drug_all[i,train_samples])
  colnames(tmp1) = "Active"
  
  tmp2 = Meta_all[train_samples,]
  tmp3 = cbind(tmp1,tmp2)
  tmp4 = tmp3[,c("Active","Time","PiD")]
  
  tmp5 = reshape(tmp4, direction = "wide", idvar = "PiD", timevar = "Time")
  tmp6 = tmp5$Active.MID-tmp5$Active.PRE
  tmp7 = tmp5$Active.POST-tmp5$Active.PRE
  names(tmp6) = tmp5$PiD
  names(tmp7) = tmp5$PiD
  
  Mid_v_Pre = rbind(Mid_v_Pre,tmp6)
  Pst_v_Pre = rbind(Pst_v_Pre,tmp7)
}

rm(list = c("tmp1","tmp2","tmp3","tmp4","tmp5","tmp6","tmp7"))

rownames(Mid_v_Pre) = row.names(Drug_all)
Mid_v_Pre = Mid_v_Pre[ , colSums(is.na(Mid_v_Pre)) == 0]
Mid_v_Pre = Mid_v_Pre[rownames(Rx_top),]

rownames(Pst_v_Pre) = row.names(Drug_all)
Pst_v_Pre = Pst_v_Pre[ , colSums(is.na(Pst_v_Pre)) == 0]
Pst_v_Pre = Pst_v_Pre[rownames(Rx_top),]

Rx_top = cbind(Rx_top,shift=rowMeans(cbind(Mid_v_Pre,Pst_v_Pre)))
Rx_top

write.csv(Rx_top,"./Reports/Drug_LmPiD_AUC_MID.csv")

thresh_Pval = 0.05
thresh_FC = 0.05

tmp = Rx_top[intersect(which(Rx_top$P.Value <= thresh_Pval),which(abs(Rx_top$shift) >= thresh_FC)),]
up = as.integer(length(which(tmp$shift > 0)))
down =  as.integer(length(which(tmp$shift <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = Rx_top, x = ~shift, y = ~-1*log10(P.Value))%>%
  layout(xaxis = list(title = "Delta", tickfont = list(size = 40),titlefont = list(size = 60)),
         yaxis = list(title = "-log(p-value)", tickfont = list(size = 40),titlefont = list(size = 60)),
         shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(Rx_top),marker = list(color='black'))

plotly::export(p=fig,file = "Reports/Drug_LmPiD_AUC_MID.png")
fig

rm('design','fig','group_df','metadata','row_dend','rAnno')
```

```{r}
panel.hist <- function(x, ...)
{
    usr <- par("usr"); on.exit(par(usr))
    par(usr = c(usr[1:2], 0, 1.5) )
    h <- hist(x, plot = FALSE)
    breaks <- h$breaks; nB <- length(breaks)
    y <- h$counts; y <- y/max(y)
    rect(breaks[-nB], 0, breaks[-1], y, col = "black", ...)
}

pairs(Drug_all[,c('PIM300','PIM328')], diag.panel=panel.hist)
```
```{r}
diff = (Drug_all$PIM300)-(Drug_all$PIM328)
TNBC271 = row.names(Drug_all)[which(abs(diff)>0.25)]
Drug_meta[TNBC271,] %>% select('Cmpd1','Class','Target')
```

```{r}
panel.hist <- function(x, ...)
{
    usr <- par("usr"); on.exit(par(usr))
    par(usr = c(usr[1:2], 0, 1.5) )
    h <- hist(x, plot = FALSE)
    breaks <- h$breaks; nB <- length(breaks)
    y <- h$counts; y <- y/max(y)
    rect(breaks[-nB], 0, breaks[-1], y, col = "black", ...)
}

pairs(Drug_all[,c('PIM388','PIM393','PIM404')], diag.panel=panel.hist)
```

```{r}
diff = (Drug_all$PIM388)-(Drug_all$PIM393)
ATID608 = row.names(Drug_all)[which(abs(diff)>0.25)]
ATID608
```

```{r}
x = list(PDX1 = ATID608, PDX2 = TNBC271)

library(ggvenn)
ggvenn(x,
       fill_color = c("#0073C2FF","#CD534CFF"),
       stroke_size = 0.5, set_name_size = 4)
rm('x','r2','test','gr')
```


```{r fig.height=40, fig.width=32}
library("ComplexHeatmap")
library("dplyr")

group_df = as.data.frame(Drug_meta[row.names(Rx_top),] %>% dplyr::select("Class"),use_raster = TRUE, raster_by_magick = TRUE)
anno_df = as.data.frame(Drug_meta[row.names(Rx_top),] %>% dplyr::select("Target"),use_raster = TRUE, raster_by_magick = TRUE)
anno_df = cbind(X1_v_X0=rowMeans(Mid_v_Pre),XS_v_X0=rowMeans(Pst_v_Pre),anno_df)
anno_txt = anno_text(anno_df$Target, which = "row")

library(circlize)
col_fun = colorRamp2(c(-0.33,0,0.33), c("blue","white","red"))
#col_fun(seq(-10, 10))

rAnno = rowAnnotation(df = anno_df,
                      col=list(X1_v_X0=col_fun,XS_v_X0=col_fun), 
                      width = ncol(anno_df)*unit(10, "mm")
                      )

hm1 = Heatmap(Mid_v_Pre[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0)),],
              width = ncol(Mid_v_Pre)*unit(20, "mm"),
              name = "ΔAUC",
              clustering_method_rows = "average",
              column_title = "MID-PRE",
              column_title_gp=gpar(fontsize=42,fontface = "bold"),
              column_names_gp = gpar(fontsize = 36),
              row_split = group_df$Class[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0))],
              gap = unit(2,"mm"),
              row_title_rot = 0,
              row_title_gp=gpar(fontsize=42,fontface = "bold"),
              heatmap_legend_param = list(
                legend_direction = "horizontal",
                legend_width = unit(6, "cm"),
                title_position = "topcenter",
                title_gp = gpar(fontsize = 36),
                labels_gp = gpar(fontsize = 36)),
              col = col_fun,
              use_raster = TRUE, raster_by_magick = TRUE)

hm2 = Heatmap(Pst_v_Pre[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0)),],
              width = ncol(Pst_v_Pre)*unit(20, "mm"),
              name = "ΔAUC",
              column_title = "POST-PRE",
              column_title_gp=gpar(fontsize=42,fontface = "bold"),
              column_names_gp = gpar(fontsize = 36),
              heatmap_legend_param = list(
                legend_direction = "horizontal",
                legend_width = unit(6, "cm"),
                title_position = "topcenter",
                title_gp = gpar(fontsize = 36),
                labels_gp = gpar(fontsize = 36)),
              row_labels = anno_df$Target[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0))],
              row_names_gp = gpar(fontsize = 30),
              col = col_fun,
              use_raster = TRUE, raster_by_magick = TRUE)

ht_list = hm1+hm2+anno_txt

draw(ht_list,
     row_title = "dAUC",
     row_title_gp = gpar(col = "black"),
     heatmap_legend_side = "bottom",
     annotation_legend_side = "right",
     legend_labels_gp = gpar(100))

#rm('anno_df','anno_txt','hm1','hm2','ht_list','Mid_v_Pre','Pst_v_Pre','rAnno','group_df')
```

```{r}
library(fgsea)
library(corto)
pathways = gmtPathways("R:\\TAMU-IBT_Powell\\Code\\R_scripts\\ssGSEA2.0-master\\db\\msigdb\\canonical.gmt")
RNA_ssgsea = ssgsea(RNA_all,pathways,scale=TRUE,minsize = 10)
colnames(RNA_ssgsea) = colnames(RNA_all)
```

Train loader
```{r}
Y_all = t(Drug_all[drug_filter,])
drugnames = colnames(Y_all)

X_all = t(RNA_all[,]) #Use COMBAT normalized values
#X_all = t(RNA_net[,]) #use Network smoothing RNAseq
#X_all = t(RNA_ssgsea[,]) #use simplied ssGSEA as the input
```

In vitro drug-gene correlation followed by gene set enrichment analysis
```{r}
library("pathfindR")
library("stringr")
library("dplyr")

tmp = cbind(Y_all[CCL_complete,"PEVONEDISTAT_Broad"],X_all[CCL_complete,])
tmp = apply(tmp[, -1], 2, function(x) {cor.test(tmp[, 1], x, method = "pearson", use = "pairwise")})

p.vals <- as.matrix(unlist(sapply(tmp, "[[", "p.value")))
r.vals <- as.matrix(unlist(sapply(tmp, "[[", "estimate")))

tmp = as.data.frame(cbind(colnames(X_all),
                          r_val = r.vals,
                          p_val = p.vals))
colnames(tmp) = c("gene_symbol","pearson","P_value")
rownames(tmp) <- NULL

tmp$pearson = as.numeric(tmp$pearson)
tmp$P_value = as.numeric(tmp$P_value)

pfr_name = "./Reports/Corr_complete_RNA_Pevonedistat_pathfindR"
unlink(pfr_name, recursive = TRUE)

pathfindR <- run_pathfindR(tmp,
                           iterations = 10,
                           gene_sets = "Reactome",
                           p_val_threshold = 0.05,
                           search_method = "GR",
                           plot_enrichment_chart = FALSE,
                           output_dir = pfr_name,
                           silent_option = TRUE)
```

In vivo drug-gene correlation followed by gene set enrichment analysis
```{r message=TRUE, warning=FALSE, paged.print=FALSE}
library("pathfindR")
library("stringr")
library("dplyr")

InVivo_idx = which(!is.na(Meta_all$InVivo_PEV))
InVivo_list = Meta_all[InVivo_idx,"InVivo_PEV"]
InVivo_models = Meta_all[InVivo_idx,"PDX_ID"]

tmp = X_all[unlist(as.matrix(InVivo_models)),]
tmp = cbind(InVivo_list,tmp)
tmp = apply(tmp[, -1], 2, function(x) {cor.test(tmp[, 1], x, method = "pearson", use = "pairwise")})
#tmp = apply(tmp[, -1], 2, function(x) {wilcox.test(tmp[, 1], x, use = "pairwise")})

p.vals <- as.matrix(unlist(sapply(tmp, "[[", "p.value")))
r.vals <- as.matrix(unlist(sapply(tmp, "[[", "estimate")))

tmp = as.data.frame(cbind(colnames(X_all),
                          r_val = r.vals,
                          p_val = p.vals))
colnames(tmp) = c("gene_symbol","pearson","P_value")
rownames(tmp) <- NULL

tmp$pearson = as.numeric(tmp$pearson)
tmp$P_value = as.numeric(tmp$P_value)

pfr_name = "./Reports/ART-TNBC_Corr_InVivoPEV_KEGG"
unlink(pfr_name, recursive = TRUE)

pathfindR <- run_pathfindR(tmp,
                               iterations = 10,
                               p_val_threshold = 0.05,
                               search_method = "GR",
                               output_dir = pfr_name,
                               silent_option = TRUE)
pathfindR

library("visNetwork")

result_df = pathfindR
result_df <- result_df[order(result_df$lowest_p, decreasing = FALSE), ]

ID_column <- "Term_Description"

edges <- data.frame()
all_up = c()
all_down = c()
for (i in base::seq_len(nrow(result_df))) {
  up_genes <- unlist(strsplit(result_df$Up_regulated[i],", "))
  all_up = unique(c(all_up,up_genes))
  down_genes <- unlist(strsplit(result_df$Down_regulated[i],", "))
  all_down = unique(c(all_down,down_genes))
  genes <- c(up_genes, down_genes)
  for (gene in genes) {
    edges <- rbind(edges, data.frame(Term = result_df[i, ID_column], Gene = gene))
  }
}

pathways = as.data.frame(unique(edges$Term))
colnames(pathways) = "id"
pathways[,"color"] = "grey"
pathways[,"group"] = "Pathway"
pathways[,"shape"] = "square"

all_up = as.data.frame(all_up)
colnames(all_up) = "id"
all_up[,"color"] = "green"
all_up[,"group"] = "Up_regulated"
all_up[,"shape"] = "circle"

all_down = as.data.frame(all_down)
colnames(all_down) = "id"
all_down[,"color"] = "darkred"
all_down[,"group"] = "Down_regulated"
all_down[,"shape"] = "circle"

nodes = rbind(pathways,all_up,all_down)

colnames(edges) = c("from","to")

rm("pathways","all_up","all_down")

graph = visNetwork(nodes, edges,width='100%',height=2000)%>%
  visPhysics(stabilization = list(iterations=500))%>%
  visNodes(color = list(inherit=FALSE), font = list(size = 30))%>%
  visEdges(color = list(inherit=FALSE, font = list(size = 40)))%>%
  visOptions(highlightNearest = list(enabled = T, degree = 2, hover = T)) %>%
  visInteraction(multiselect = TRUE) %>%
  visLayout(randomSeed = 123)%>%
  visGroups(groupname = "Pathway", color = "grey") %>%
  visGroups(groupname = "Up_regulated", color = "green") %>%
  visGroups(groupname = "Down_regulated", color = "darkred") %>%
  visLegend(position = "right")

visSave(graph, file=paste(pfr_name,"visNetwork.html", sep="/"), selfcontained = TRUE, background = "white")

InVivo_Pev_path = pathfindR
Invivo_Pev_genes = tmp
write.csv(Invivo_Pev_genes,file = "./Reports/Primary/Invivo_Pev_genes_pearson.csv")
```

```{r}
input_processed = tmp$pearson
names(input_processed) = tmp$gene_symbol

hsa_preview = function(pathfindR,input_df,outdir="./hsa_preview"){
  
  #pathview_in = input_processed #input_df %>% dplyr::select("CHANGE")
  
  for(i in 1:nrow(pathfindR)){
    mypathway<-pathfindR$ID[i]
    Pathway_name = pathfindR$Term_Description[i]
    dir.create(file.path(outdir, Pathway_name), showWarnings = FALSE,recursive = TRUE)
    setwd(file.path(outdir, Pathway_name))
    
    library(pathview)
    pathview(gene.data=input_df,
             species="hsa",
             pathway=mypathway,gene.idtype = "SYMBOL",
             low = list(gene = "red", cpd = "blue"),
             mid = list(gene = "gray", cpd = "gray"), 
             high = list(gene = "green", cpd = "yellow"))
  }
}
row.names(Invivo_Pev_genes) = Invivo_Pev_genes$gene_symbol
Invivo_Pev_genes = Invivo_Pev_genes[which(Invivo_Pev_genes$P_value <= 0.05),]

hsa_preview(pathfindR,input_processed)

#visualize_hsa_KEGG(pathfindR$ID,input_processed)
```

Tool to spot check/export in vitro correlative data
```{r}
df = as.data.frame(t(rbind(Resp = Drug_all["PEVONEDISTAT_Broad",train_samples],RNA_all[c("NAE1","SAE1"),train_samples])))

library(ggplot2)
library(ggpubr)

# Create a scatter plot with a linear trend line and correlation statistic overlay
ggplot(df, aes(x = Resp, y = SAE1)) +
  geom_point() +  # Add scatter points
  geom_smooth(method = "lm", se = FALSE, color = "blue") +  # Add linear trend line
  stat_cor(method = "pearson", label.x = 0, label.y = 0, size = 4) +  # Add correlation statistic

# Customize the plot further if needed
  labs(
    title = "Scatter Plot with Linear Trend Line and Correlation Statistic",
    x = "PEVONEDISTAT (Normalized Response)",
    y = "NAE1"
  ) +
  theme_minimal()
```
Tool to spot check/export in vivo correlative data
```{r}
df = as.data.frame(cbind(Resp=as.matrix(Meta_all[train_samples,"InVivo_PEV"]),t(RNA_all[c("NAE1","SAE1","SUMO1"),train_samples])))
df = df[!is.na(df$InVivo_PEV),]

library(ggplot2)
library(ggpubr)

# Create a scatter plot with a linear trend line and correlation statistic overlay
ggplot(df, aes(x = InVivo_PEV, y = SAE1)) +
  geom_point() +  # Add scatter points
  geom_smooth(method = "lm", se = FALSE, color = "blue") +  # Add linear trend line
  stat_cor(method = "pearson", label.x = 0, label.y = 0, size = 4) +  # Add correlation statistic

# Customize the plot further if needed
  labs(
    title = "Scatter Plot with Linear Trend Line and Correlation Statistic",
    x = "PEVONEDISTAT (Normalized Response)",
    y = "NAE1"
  ) +
  theme_minimal()
```






```{r}
library("caret")
library("doParallel")
#cl <- makePSOCKcluster(30)
#registerDoParallel(cl)


Result_list = c()

for (active_drug_index in 1:ncol(Y_all)){
  Cmpd_results = c()
  active_drug_name = drugnames[active_drug_index]
  
  #for (geneSet in 1:10){
  for (geneSet in 1:length(pathways)){
    model_tmp = c()
      
    activeSet = intersect(row.names(RNA_all),pathways[[geneSet]])
    active_pathway_name = names(pathways)[geneSet]
    
    control=trainControl(method = "LOOCV",number=33)
    lasso <- train(X_all[train_samples,activeSet],
                   Y_all[train_samples,active_drug_index], 
                   method = "glmnet", 
                   lambda= 0, 
                   tuneGrid = expand.grid(alpha = 1,  lambda = 0),
                   trControl=control)
    
    #model_tmp[activeName] = lasso[["results"]][["RMSE"]]
    Cmpd_results = rbind(Cmpd_results,lasso[["results"]][["RMSE"]])
    row.names(Cmpd_results)[geneSet] = names(pathways)[geneSet]
  }
  Result_list[[active_drug_name]] = Cmpd_results
}

#stopCluster(cl)

```

CV
```{r}
library(stringr)
library(sets)
library(caret)

Cmpd_results = c()
CCL_complete = intersect(row.names(X_all),row.names(Y_all))
active_drug_index = which(str_detect(drugnames,"PEVONEDISTAT"))
active_drug_name = drugnames[active_drug_index]

#for (geneSet in 1:10){
for (geneSet in 1:length(pathways)) {
  try({
        model_tmp = c()
        
        activeSet = intersect(colnames(X_all),pathways[[geneSet]])
        active_pathway_name = names(pathways)[geneSet]
        
        control=trainControl(method = "LOOCV",
                             number=34,
                             allowParallel = TRUE)
        lasso <- train(X_all[CCL_complete,activeSet],
                       Y_all[CCL_complete,active_drug_index],
                       method = "glmnet",
                       lambda= 0,
                       tuneGrid = expand.grid(alpha = 1,  lambda = 0),
                       trControl=control)
        
        importance = varImp(lasso)
          
        #model_tmp[activeName] = lasso[["results"]][["RMSE"]]
        model_tmp['pathway'] = names(pathways)[geneSet]
        model_tmp['RMSE'] = lasso[["results"]][["RMSE"]][1]
        model_tmp["Rsq"] = lasso[["results"]][["Rsquared"]]
        model_tmp["Pearson"] = cor(Y_all[CCL_complete,active_drug_index],lasso[["pred"]][["pred"]],method="pearson")
        model_tmp["impNames"] = paste(row.names(importance[["importance"]]),collapse = ";")
        model_tmp["impValues"] = paste(importance[["importance"]][["Overall"]],collapse = ";")
        model_tmp["Sample"] = paste(names(Y_all[CCL_complete,active_drug_index]),collapse = ";")
        model_tmp["actual"] = paste(Y_all[CCL_complete,active_drug_index],collapse = ";")
        model_tmp["predicted"] = paste(lasso[["pred"]][["pred"]],collapse = ";")
        
        Cmpd_results = rbind(Cmpd_results,model_tmp)
        
      }, silent = FALSE, outFile = getOption("try.outFile", default = stderr()))
          
}

row.names(Cmpd_results) = NULL
```

```{r}
write.csv(Cmpd_results,'./Reports/ART-TNBC_Pevonedistat_CV.csv')
```

Trains a serires of L1 regularized (lasso) regression models. Each model is trained against a single gene set and the RMSE is used to rank the importance of individual pathways. Feature importance analysis provide a more grainular view of the individual genes that are driving the decision making.  
```{r}
library(stringr)
library(foreach)
library(doParallel)
library(caret)

Query = 'PEVONEDISTAT'

y = Y_all[,which(str_detect(colnames(Y_all),Query))]
names(y) = row.names(Y_all)
y = y[!is.na(y)]
subset = names(y)

x = X_all
# Number of bootstrap samples
num_bootstrap_samples <- 10
n = floor(0.9*dim(x)[1])

# Register parallel backend (adjust the number of cores as needed)
# This example uses parallel execution on 4 cores, but you can change it.
cl <- makeCluster(10)
registerDoParallel(cl)

# Storage for cross-validation results
cv_results <- foreach(i = 1:num_bootstrap_samples, .packages = c('glmnet','caret')) %dopar% {
      set.seed(i)
      cv_result = list()
      #select by geneSet
      for(j in names(pathways)){
        try({
              #init
              tmp =c()
              
              #Select Geneset using msigdb
              geneSet = intersect(colnames(x),pathways[[j]])
              
              # Create a bootstrap sample
              sample_indices <- sample(1:n, replace = TRUE)
              X_bootstrap <- x[sample_indices,geneSet]
              y_bootstrap <- y[sample_indices]
             
              # Fit a glmnet model with LOOCV
              control=trainControl(method = "LOOCV",
                                   number=n,
                                   allowParallel = FALSE)
              
              lasso <- train(X_bootstrap,
                             y_bootstrap,
                             method = "glmnet",
                             lambda= 0,
                             tuneGrid = expand.grid(alpha = 1,  lambda = 0),
                             trControl=control)
              
              
               importance = varImp(lasso)
                
               tmp["RMSE"] = lasso[["results"]][["RMSE"]]
               tmp["Rsq"] = lasso[["results"]][["Rsquared"]]
               tmp["Pearson"] = cor(y_bootstrap,lasso[["pred"]][["pred"]],method="pearson")
               tmp["impNames"] = paste(row.names(importance[["importance"]]),collapse = ";")
               tmp["impValues"] = paste(importance[["importance"]][["Overall"]],collapse = ";")
               tmp["Sample"] = paste(names(y_bootstrap),collapse = ";")
               tmp["actual"] = paste(y_bootstrap,collapse = ";")
               tmp["predicted"] = paste(lasso[["pred"]][["pred"]],collapse = ";")
               cv_result[[j]] = tmp
            }, silent = FALSE, outFile = getOption("try.outFile", default = stderr()))
      }
      do.call(rbind, cv_result)
}

# Stop the parallel cluster
stopCluster(cl)
```

```{r}
df = as.data.frame(do.call(rbind, cv_results))
df[,"group"] = as.factor(row.names(do.call(rbind, cv_results)))
rownames(df) = NULL
write.csv(df,'./Reports/ART-TNBC_Pevonedistat_BootCV.csv')
```


```{r}

library("doParallel")

cores=detectCores()
cl <- makeCluster(cores[1]-2) #not to overload your computer
registerDoParallel(cl)


Result_list = c()

for (active_drug_index in 1:5){
  Cmpd_results = c()
  active_drug_name = drugnames[active_drug_index]
  
  #for (geneSet in 1:10){
  Result_list[[active_drug_name]] = foreach (i=1:1000,.combine=rbind) %dopar% {
    
    library("caret")
    model_tmp = c()
      
    activeSet = intersect(row.names(RNA_all),pathways[[geneSet]])
    active_pathway_name = names(pathways)[geneSet]
    
    control=trainControl(method = "cv",number=5)
    lasso <- train(X_all[train_samples,activeSet],
                   Y_all[train_samples,active_drug_index], 
                   method = "glmnet", 
                   lambda= 0, 
                   tuneGrid = expand.grid(alpha = 1,  lambda = 0),
                   trControl=control)
    
    #model_tmp[activeName] = lasso[["results"]][["RMSE"]]
    Cmpd_results = rbind(Cmpd_results,lasso[["results"]][["RMSE"]])
    row.names(Cmpd_results)[geneSet] = names(pathways)[geneSet]
  }
}

stopCluster(cl)

```









Next, we evaluated the similarity of the matched pairs to determine whether individual patients or time points tended to associate with one another. From this analysis we found that individual patients tended to co-cluster over time points. Thus recapitulating a strong degree of patient heterogeneity.

```{r echo=FALSE}
library("dendextend")
library("ComplexHeatmap")
library("dplyr")

row_variances <- apply(RNA_all[,train_samples], 1, var)
sorted_indices <- order(row_variances,decreasing = T)
RNA_all_sort <- RNA_all[sorted_indices, ]
RNA_all_sort[1:250,]

CCL_corr = cor(as.matrix(RNA_all_sort[1:250,train_samples]),method = c("pearson"))
hc = hclust(dist(t(CCL_corr),method = "euclidean"),method = "complete")
df_anno = Meta_all[train_samples,] %>% dplyr::select("Time","PiD")


row_dend = as.dendrogram(hc)

library(circlize)
col_fun = colorRamp2(c(-1, 0, 1), c("blue", "white", "red"))
col_fun(seq(-3, 3))

HM = Heatmap(CCL_corr,
             name = "Pearson",
             col = col_fun,
             cluster_rows = row_dend,
             cluster_columns = row_dend,
             top_annotation = anno_gen(df_anno),
             show_row_names = FALSE)
HM
rm('HM','CCL_corr','hc','df_anno','row_dend')
```

We then perform differential gene expression analysis that accounted for the individual patient matched pairs. 
```{r warning=FALSE}
library("edgeR")
library("plotly")

metadata = Meta_all[train_samples,]
metadata[] <- lapply(metadata, factor)

design = model.matrix(~0+Time+PiD,metadata)
DEG_fit = lmFit(RNA_filtered_z[,train_samples],design)
cm = makeContrasts(Mid_v_Pre=TimeMID-TimePRE,levels=design)
DEG_fit_cm = contrasts.fit(DEG_fit,cm)
DEG_fit_cm_eBayesfit = eBayes(DEG_fit_cm,trend = FALSE,robust = FALSE)
DEG_top = topTable(DEG_fit_cm_eBayesfit,adjust.method="fdr", number=Inf)

# DEG_fit_cm_eBayesfit = eBayes(DEG_fit,trend = FALSE,robust = FALSE)
# DEG_top = topTable(DEG_fit_cm_eBayesfit, coef = "TimePRE",adjust.method="fdr", number=Inf)

DEG_top
#write.csv(DEG_top,"./Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre.csv")

thresh_Pval = 0.05
thresh_FC = 0

tmp = DEG_top[intersect(which(DEG_top$P.Value <= thresh_Pval),which(abs(DEG_top$logFC) >= thresh_FC)),]
up = as.integer(length(which(tmp$logFC > 0)))
down =  as.integer(length(which(tmp$logFC <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = DEG_top, x = ~logFC, y = ~-1*log10(P.Value),color = ~logFC, colors="RdBu")%>%
  layout(xaxis = list(title = "DeltaZ", tickfont = list(size = 40),titlefont = list(size = 60)),
         yaxis = list(title = "-log(p-value)", tickfont = list(size = 40),titlefont = list(size = 60)),
         shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(DEG_top))

#plotly::export(p=fig,file = "Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre.png")
  
fig
rm('fig','cm','design','metadata')
```

This section does the same DEG with out using repreated samples analysis
```{r warning=FALSE}
library("edgeR")
library("plotly")

metadata = Meta_all[train_samples,]
metadata[] <- lapply(metadata, factor)

design = model.matrix(~0+Time,metadata)
DEG_fit = lmFit(RNA_filtered_z[,train_samples],design)
cm = makeContrasts(Mid_v_Pre=TimeMID-TimePRE,levels=design)
DEG_fit_cm = contrasts.fit(DEG_fit,cm)
DEG_fit_cm_eBayesfit = eBayes(DEG_fit_cm,trend = FALSE,robust = FALSE)
DEG_top = topTable(DEG_fit_cm_eBayesfit,adjust.method="fdr", number=Inf)

# DEG_fit_cm_eBayesfit = eBayes(DEG_fit,trend = FALSE,robust = FALSE)
# DEG_top = topTable(DEG_fit_cm_eBayesfit, coef = "TimePRE",adjust.method="fdr", number=Inf)

DEG_top
#write.csv(DEG_top,"./Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre.csv")

thresh_Pval = 0.05
thresh_FC = 0

tmp = DEG_top[intersect(which(DEG_top$P.Value <= thresh_Pval),which(abs(DEG_top$logFC) >= thresh_FC)),]
up = as.integer(length(which(tmp$logFC > 0)))
down =  as.integer(length(which(tmp$logFC <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = DEG_top, x = ~logFC, y = ~-1*log10(P.Value),color = ~logFC, colors="RdBu")%>%
  layout(xaxis = list(title = "DeltaZ", tickfont = list(size = 40),titlefont = list(size = 60)),
         yaxis = list(title = "-log(p-value)", tickfont = list(size = 40),titlefont = list(size = 60)),
         shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(DEG_top))

#plotly::export(p=fig,file = "Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre.png")
  
fig
rm('fig','cm','design','metadata')
```

Gene-pathway enrichment analysis using the DGE list from the longitudinal matched pairs.
```{r message=TRUE, warning=FALSE}
library("pathfindR")
library("visNetwork")

pfR_input = DEG_top[,c("logFC","P.Value"),]
Gene_symbol = row.names(pfR_input)
pfR_input = cbind(Gene_symbol,pfR_input)

name = "Reports/DEG_LmPiD_RNA_PDXmid_v_PDXpre_pathfindR"

#Delete old
unlink(name, recursive = TRUE)

pfr_tmp <- run_pathfindR(pfR_input,
                               iterations = 10,
                               p_val_threshold = 0.05,
                               search_method = "GR",
                               plot_enrichment_chart = FALSE,
                               output_dir = name,
                               silent_option = TRUE)

pfr_tmp <- pfr_tmp[order(pfr_tmp$lowest_p, decreasing = FALSE), ]
assign(paste("pfr_",name,sep=""),pfr_tmp)

pathfindR_visNetwork(pfr_tmp,paste(name,"visNetwork.html",sep="/"))
hsa_preview(pfr_tmp,pfR_input,outdir = paste(Base_dir,name,"hsa_preview",sep="/"))
xlsx::write.xlsx(pfr_tmp,file = paste(Base_dir,name,"pathfindR_enrichment.xlsx",sep="/"),sheetName='pathfindR')

rm("pfr_tmp")
```

```{r}
save.image(file='TNBC-PGx_analysis.RData')
sessionInfo()
```
