---
title: "Untitled"
author: "Reid T. Powell"
date: "2023-01-24"
output: html_document
---
plotly annotation tools
```{r}
vline <- function(x = 0, color = "black") {
  list(
    type = "line",
    y0 = 0,
    y1 = 1,
    yref = "paper",
    x0 = x,
    x1 = x,
    line = list(color = color, dash="dot"))

}

hline <- function(y = 0, color = "black") {
  list(
    type = "line",
    x0 = 0,
    x1 = 1,
    xref = "paper",
    y0 = y,
    y1 = y,
    line = list(color = color, dash="dot"))
}

anno_gen = function(df){
  library(ComplexHeatmap)
  hm_anno = HeatmapAnnotation(df = as.data.frame(df),
                            col=list(
                              PDX_ = c("PIM025"="darksalmon","PIM038"="darkslateblue","PIM254"="darkturquoise","PIM311"="goldenrod4"),
                              Class_ = c("Non_responsive"="red4","Responsive"="springgreen4"),
                              Time_ = c("D00"="grey0","D04"="grey40","D07"="grey80","D14"="grey99"),
                              Rx_ = c("VEH"="dodgerblue4","PEV"="chartreuse4")
                              )
                            )
  return(hm_anno)
}

pathfindR_visNetwork = function(pfr_tmp,flnm="visNetwork.html"){
  ID_column <- "Term_Description"

  edges <- data.frame()
  all_up = c()
  all_down = c()
  for (i in base::seq_len(nrow(pfr_tmp))) {
    up_genes <- unlist(strsplit(pfr_tmp$Up_regulated[i],", "))
    all_up = unique(c(all_up,up_genes))
    down_genes <- unlist(strsplit(pfr_tmp$Down_regulated[i],", "))
    all_down = unique(c(all_down,down_genes))
    genes <- c(up_genes, down_genes)
    for (gene in genes) {
      edges <- rbind(edges, data.frame(Term = pfr_tmp[i, ID_column], Gene = gene))
    }
  }
  
  pathways = as.data.frame(unique(edges$Term))
  colnames(pathways) = "id"
  pathways[,"color"] = "grey"
  pathways[,"group"] = "Pathway"
  pathways[,"shape"] = "square"
  pathways[,"font.size"] = 30
  
  all_up = as.data.frame(all_up)
  colnames(all_up) = "id"
  all_up[,"color"] = "green"
  all_up[,"group"] = "Up_regulated"
  all_up[,"shape"] = "circle"
  all_up[,"font.size"] = 20
  
  all_down = as.data.frame(all_down)
  colnames(all_down) = "id"
  all_down[,"color"] = "darkred"
  all_down[,"group"] = "Down_regulated"
  all_down[,"shape"] = "circle"
  all_down[,"font.size"] = 20
  
  nodes = rbind(pathways,all_up,all_down)
  
  colnames(edges) = c("from","to")
  
  rm("pathways","all_up","all_down")
  
  graph = visNetwork(nodes, edges,width='100%',height=2000)%>%
    visPhysics(stabilization = list(iterations=500))%>%
    visNodes(color = list(inherit=FALSE), font = list(size = 30))%>%
    visEdges(color = list(inherit=FALSE, font = list(size = 40)))%>%
    visOptions(highlightNearest = list(enabled = T, degree = 2, hover = T)) %>%
    visInteraction(multiselect = TRUE) %>%
    visLayout(randomSeed = 123)%>%
    visGroups(groupname = "Pathway", color = "grey") %>%
    visGroups(groupname = "Up_regulated", color = "green") %>%
    visGroups(groupname = "Down_regulated", color = "darkred") %>%
    visLegend(position = "right")
  
  visSave(graph, file=flnm, selfcontained = TRUE, background = "white")
}

hsa_preview = function(pfr_results,input_df,outdir="./hsa_preview"){
  
  pathview_in = pfR_input %>% dplyr::select("logFC")
  
  for(i in 1:nrow(pfr_results)){
    mypathway<-pfr_results$ID[i]
    Pathway_name = pfr_results$Term_Description[i]
    dir.create(file.path(outdir, Pathway_name), showWarnings = FALSE,recursive = TRUE)
    setwd(file.path(outdir, Pathway_name))
    library(pathview)
    pathview(gene.data=pathview_in,
             species="hsa",
             pathway=mypathway,gene.idtype = "SYMBOL",
             low = list(gene = "red", cpd = "blue"),
             mid = list(gene = "gray", cpd = "gray"), 
             high = list(gene = "green", cpd = "yellow"))
  }
}
```


Read in the experimental RNAseq data from the rseqR analysis
```{r warning=FALSE}
#BiocManager::install("EnsDb.Hsapiens.v79") ##Install
library("readxl")
library("EnsDb.Hsapiens.v79")

Base_dir = "D:\\Users\\rpowell\\Documents\\GitHub\\TNBC-PGx"
knitr::opts_knit$set(root.dir = Base_dir)
setwd(Base_dir)

Meta_all <- read_excel("./data2/MASTER.xlsx",col_names=TRUE)
Meta_all[] <- lapply(Meta_all, factor)

RNA_all <- read.csv("./data2/counts.csv", row.names=1)
RNA_all = RNA_all[,Meta_all$Sample]
```

Before differential gene expression analysis is performed low expression reads are removed. The trimmed mean of M (TMM) values is then calculated and applied as a correction factor to adjust for large shifts in global expression. Finally the data are normalized to the library size, as counts per million (CPM), and log normalized.
```{r message=FALSE, warning=FALSE, paged.print=FALSE}
library("RColorBrewer")
library("edgeR")

RNA_exp = DGEList(RNA_all,remove.zeros	= TRUE)

#calculate and apply normalization factor
RNA_exp <- calcNormFactors(RNA_exp, method = "TMM")
RNA_exp_cpm = cpm(RNA_exp, log=TRUE, normalized.lib.sizes = TRUE)

row.means = rowMeans(RNA_exp_cpm)
keep = which(row.means >= 0)

RNA_exp_cpm_filter = RNA_exp_cpm[keep,]
RNA_exp_cpm_filter_z = t(scale(t(RNA_exp_cpm_filter),center = T, scale = T)) #z-transformed network 

nsamples <- ncol(RNA_exp_cpm)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(as.matrix(RNA_exp_cpm[,1])), col=col[1], lwd=2, ylim=c(0,1), las=2, main="", xlab="")
title(main="Log2 CPM (ALL)", xlab="Log-cpm")
#abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
	den <- density(as.matrix(RNA_exp_cpm[,i]))
	lines(den$x, den$y, col=col[i], lwd=2)
}

plot(density(as.matrix(RNA_exp_cpm_filter[,1])), col=col[1], lwd=2, ylim=c(0,1), las=2, main="", xlab="")
title(main="Log2 CPM (Filtered)", xlab="Log-cpm")
#abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
	den <- density(as.matrix(RNA_exp_cpm_filter[,i]))
	lines(den$x, den$y, col=col[i], lwd=2)
}
rm("den")
```
Correlation network of the z-normalized gene expression.
```{r message=FALSE, warning=FALSE}
library("ComplexHeatmap")
library("dplyr")
library("dendextend")

CCL_corr = cor(RNA_exp_cpm_filter_z,method = c("pearson"))

plot_hm_row_dend = as.dendrogram(hclust(dist(CCL_corr)))
plot_hm_row_dend = color_branches(plot_hm_row_dend, k = 6)
                            
plot_hm = Heatmap(CCL_corr,
             name = "Pearson",
             cluster_rows = plot_hm_row_dend,
             cluster_columns = plot_hm_row_dend,
             top_annotation = anno_gen(Meta_all %>% dplyr::select("Class_","PDX_","Time_","Rx_")),
             show_row_names = FALSE,
             show_column_names = FALSE)
plot_hm
rm("CCL_corr")
```

```{r message=FALSE, warning=FALSE, paged.print=FALSE}
library(factoextra)
library(plotly)

plot_pca = prcomp(t(RNA_exp_cpm_filter))
plot_pca_var = get_eig(plot_pca)
#fviz_eig(plot_pca)

plot_df = as.data.frame(plot_pca[["x"]])
plot_DGE <- plot_ly(x = plot_df$PC1, y = plot_df$PC2 , color=Meta_all$PDX_)%>%
  layout(xaxis = list(title = paste("PC1 (",format(plot_pca_var$variance.percent[1],digits=4),"%)",sep="")),
         yaxis = list(title = paste("PC2 (",format(plot_pca_var$variance.percent[2],digits=4),"%)",sep="")))%>%
  add_trace(text=row.names(plot_df))
plot_DGE
```
corto: a lightweight R package for gene network inference and master regulator analysis. This section perfroms a fast ssGSEA analysis on the cleaned up gene expression matrix.
```{r}
library(fgsea)
library(corto)

corto.inmat = RNA_exp_cpm_filter

corto.groups = gmtPathways("./data2/canonical.gmt")
corto.ssgsea = ssgsea(corto.inmat,corto.groups)
colnames(corto.ssgsea) = colnames(corto.inmat)
```

QUESTION_1: What defines PEV-responsive versus PEV-non-responsive PDX models at the baseline.
```{r}
Cohort = Reduce(intersect,list(which(Meta_all$Time_ == "D00"),which(Meta_all$Rx_ == "VEH")))
Meta_baseline = Meta_all[Cohort,]
RNA_baseline = RNA_exp_cpm_filter[,Cohort]
ssGSEA_baseline = corto.ssgsea[,Cohort]
```

Differential pathway activation analysis
```{r fig.height=20, fig.width=24, message=FALSE, warning=FALSE}
library("edgeR")
library("plotly")

DPA_design = model.matrix(~0+Class_+PDX_,Meta_baseline)
DPA_fit = lmFit(ssGSEA_baseline,DPA_design)
DPA_cm = makeContrasts(matrix=Class_Non_responsive-Class_Responsive,levels=DPA_design)
DPA_fit_cm = contrasts.fit(DPA_fit,DPA_cm)
DPA_fit_cm_eBayesfit = eBayes(DPA_fit_cm,trend = FALSE,robust = FALSE)
DPA_top = topTable(DPA_fit_cm_eBayesfit,adjust.method="fdr", number=Inf)

DPA_top


thresh_Pval = 0.05
thresh_FC = 5

tmp = DPA_top[intersect(which(DPA_top$adj.P.Val <= thresh_Pval),which(abs(DPA_top$logFC) >= thresh_FC)),]
up = as.integer(length(which(tmp$logFC > 0)))
down =  as.integer(length(which(tmp$logFC <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = DPA_top, x = ~logFC, y = ~-1*log10(adj.P.Val),color = ~logFC, colors="RdBu")%>%
  layout(xaxis = list(title = "DeltaZ"),yaxis = list(title = "-log(p-value)"), shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(DPA_top))
fig

library(ComplexHeatmap)
DPA_select_n = 100
DPA_top_select = DPA_top[order(DPA_top$logFC),]
DPA_top_select = DPA_top_select[DPA_top_select$adj.P.Val <= 0.05,]
DPA_top_n_names = row.names(DPA_top_select[c(1:floor(DPA_select_n/2),(nrow(DPA_top_select)-floor(DPA_select_n/2)+1):nrow(DPA_top_select)),])

hm_input = ssGSEA_baseline[DPA_top_n_names,]

plot_hm = Heatmap(hm_input,
             name = "ssGSEA",
             ##top_annotation = hm_anno_base,
             show_row_names = TRUE,
             row_names_gp = gpar(fontsize = 10),
             row_names_max_width = max_text_width(rownames(hm_input),gp = gpar(fontsize = 10)),
             show_column_names = FALSE,
             heatmap_legend_param = list(
               legend_direction = "horizontal",
               legend_width = unit(6, "cm"))
             )
draw(plot_hm, heatmap_legend_side="bottom", annotation_legend_side="left",
           legend_grouping = "adjusted")
```

QUESTION_2: What changes as a function of time with PEV exposure
```{r fig.height=20, fig.width=24}
library(sigFeature)
library(dplyr)
thresh_Pval = 0.0001
search_terms = c("PIM254","PIM311")
Cohort_Drug = Reduce(intersect,list(which(Meta_all$PDX_ %in% search_terms),which(Meta_all$Rx_=="PEV")))
Cohort_VEH = Reduce(intersect,list(which(Meta_all$PDX_ %in% search_terms),which(Meta_all$Rx_=="VEH")))

SigFeat_Y_drug = Meta_all[Cohort_Drug,] %>% dplyr::select("Time_")
SigFeat_X_drug = t(corto.ssgsea[,Cohort_Drug])
SigFeat_X_drug = SigFeat_X_drug[,order(apply(SigFeat_X_drug,2,sd),decreasing = T)]
#SigFeat_X_drug = SigFeat_X_drug[,1:1000]
SigFeat_X_drug_names = colnames(SigFeat_X_drug)
sigFeat_drug = unlist(sigFeaturePvalue(x=SigFeat_X_drug,y=SigFeat_Y_drug))
names(sigFeat_drug) = SigFeat_X_drug_names
sigFeat_drug_sig = sigFeat_drug[which(sigFeat_drug<thresh_Pval)]

SigFeat_Y_VEH = Meta_all[Cohort_VEH,] %>% dplyr::select("Time_")
SigFeat_X_VEH = t(corto.ssgsea[SigFeat_X_drug_names,Cohort_VEH])
SigFeat_X_VEH_names = colnames(SigFeat_X_VEH)
sigFeat_VEH = unlist(sigFeaturePvalue(x=SigFeat_X_VEH,y=SigFeat_Y_VEH))
names(sigFeat_VEH) = SigFeat_X_VEH_names
sigGene_VEH_sig = sigFeat_VEH[which(sigFeat_VEH<0.05)]

sigGene = names(sigFeat_drug_sig)[which(!(names(sigFeat_drug_sig) %in% (sigGene_VEH_sig)))]
```

look at genes in pathways
```{r}
library(plotly)
df = as.data.frame(cbind(sigFeat_drug,sigFeat_VEH))
df = -log10(df[,])
df[,"Contrast"] = df$sigFeat_drug-df$sigFeat_VEH
df[,"Significant"] = df$Contrast >= 1.3
df = df[order(df$Contrast,decreasing = T),]
#df = df[which(str_detect(row.names(df),"HALLMARK")),]
df.names = row.names(df)

#Volcano plot of DEG ,color = ~logFC, colors="RdBu"
plot_DGE <- plot_ly(data = df, x = ~sigFeat_drug, y = ~sigFeat_VEH)%>%
  layout(xaxis = list(title = "Drug"),yaxis = list(title = "Veh"), shapes=list(vline(2),hline(1.3))) %>%
  add_trace(text=row.names(df))
plot_DGE
```

```{r fig.height=10, fig.width=16}
df.top = df[1:25,]
df.top.names = row.names(df.top)
hm_input = corto.ssgsea[df.top.names,]
#row.names(hm_input)[order(row.names(hm_input))]

HM = Heatmap(hm_input,
             name = "ssGSEA",
             top_annotation = anno_gen(Meta_all[,] %>% dplyr::select("Class_","PDX_","Time_","Rx_")),
             #right_annotation = rowAnnotation(df=as.data.frame(df$Significant[sigGene])),
             show_column_names = F,
             cluster_columns = F,
             column_split = Meta_all$PDX_,
             column_gap = unit(5, "mm"),
             column_title = NULL,
             show_row_names = TRUE,
             row_names_gp = gpar(fontsize = 15),
             row_names_max_width = max_text_width(rownames(hm_input),gp = gpar(fontsize = 6)),
             heatmap_legend_param = list(legend_direction = "horizontal", legend_width = unit(6, "cm"), title_position = "topcenter")
             )
draw(HM, heatmap_legend_side="bottom", annotation_legend_side="left",
           legend_grouping = "adjusted")
```

```{r}
sessionInfo()
```

