---
title: "TNBC Pharmacogenomic analysis"
---
Load basic libraries that are used throughout the notebook
```{r message=FALSE, warning=FALSE, include=FALSE}
rm(list=ls())
library("glmnet")
library("doMC")
library("reticulate")
library("plotly")
library("matrixStats")
library("ComplexHeatmap")
library("seriation")
library("dplyr")
library("glmnet")
library("Hmisc")
```

plotly annotation tools
```{r}
vline <- function(x = 0, color = "black") {
  list(
    type = "line",
    y0 = 0,
    y1 = 1,
    yref = "paper",
    x0 = x,
    x1 = x,
    line = list(color = color, dash="dot"))

}

hline <- function(y = 0, color = "black") {
  list(
    type = "line",
    x0 = 0,
    x1 = 1,
    xref = "paper",
    y0 = y,
    y1 = y,
    line = list(color = color, dash="dot"))
}

anno_gen = function(df){
  library(ComplexHeatmap)
hm_anno = HeatmapAnnotation(df = as.data.frame(df_anno),
                            annotation_name_side = "left",
                            col=list(
                              Time=c("PRE"="grey80",
                                     "MID"="grey40",
                                     "POST"="grey0"
                                          ),
                              PiD =c("2"="yellow3",
                                     "3"="wheat3",
                                     "11"="violetred3",
                                     "14"="turquoise3",
                                     "16"="tomato3",
                                     "18"="thistle3",
                                     "19"="tan3",
                                     "20"="steelblue3",
                                     "21"="springgreen3",
                                     "22"="snow3"
                                     )
                              )
                            )
  return(hm_anno)
}

pathfindR_visNetwork = function(pfr_tmp,flnm="visNetwork.html"){
  ID_column <- "Term_Description"

  edges <- data.frame()
  all_up = c()
  all_down = c()
  for (i in base::seq_len(nrow(pfr_tmp))) {
    up_genes <- unlist(strsplit(pfr_tmp$Up_regulated[i],", "))
    all_up = unique(c(all_up,up_genes))
    down_genes <- unlist(strsplit(pfr_tmp$Down_regulated[i],", "))
    all_down = unique(c(all_down,down_genes))
    genes <- c(up_genes, down_genes)
    for (gene in genes) {
      edges <- rbind(edges, data.frame(Term = pfr_tmp[i, ID_column], Gene = gene))
    }
  }
  
  pathways = as.data.frame(unique(edges$Term))
  colnames(pathways) = "id"
  pathways[,"color"] = "grey"
  pathways[,"group"] = "Pathway"
  pathways[,"shape"] = "square"
  pathways[,"font.size"] = 30
  
  all_up = as.data.frame(all_up)
  colnames(all_up) = "id"
  all_up[,"color"] = "green"
  all_up[,"group"] = "Up_regulated"
  all_up[,"shape"] = "circle"
  all_up[,"font.size"] = 20
  
  all_down = as.data.frame(all_down)
  colnames(all_down) = "id"
  all_down[,"color"] = "darkred"
  all_down[,"group"] = "Down_regulated"
  all_down[,"shape"] = "circle"
  all_down[,"font.size"] = 20
  
  nodes = rbind(pathways,all_up,all_down)
  
  colnames(edges) = c("from","to")
  
  rm("pathways","all_up","all_down")
  
  graph = visNetwork(nodes, edges,width='100%',height=2000)%>%
    visPhysics(stabilization = list(iterations=500))%>%
    visNodes(color = list(inherit=FALSE), font = list(size = 30))%>%
    visEdges(color = list(inherit=FALSE, font = list(size = 40)))%>%
    visOptions(highlightNearest = list(enabled = T, degree = 2, hover = T)) %>%
    visInteraction(multiselect = TRUE) %>%
    visLayout(randomSeed = 123)%>%
    visGroups(groupname = "Pathway", color = "grey") %>%
    visGroups(groupname = "Up_regulated", color = "green") %>%
    visGroups(groupname = "Down_regulated", color = "darkred") %>%
    visLegend(position = "right")
  
  visSave(graph, file=flnm, selfcontained = TRUE, background = "white")
}

hsa_preview = function(pfr_results, input_df_tmp ,outdir="./hsa_preview"){
  
  pathview_in = input_df_tmp %>% dplyr::select("LogFC")
  
  for(i in 1:nrow(pfr_results)){
    tryCatch({
    mypathway<-pfr_results$ID[i]
    Pathway_name = pfr_results$Term_Description[i]
    dir.create(file.path(outdir, Pathway_name), showWarnings = FALSE,recursive = TRUE)
    setwd(file.path(outdir, Pathway_name))
    library(pathview)
    pathview(gene.data=pathview_in,
             species="hsa",
             pathway=mypathway,gene.idtype = "SYMBOL",
             low = list(gene = "red", cpd = "blue"),
             mid = list(gene = "gray", cpd = "gray"), 
             high = list(gene = "green", cpd = "yellow"))},error=function(e){})
  }
}
```



Set working directory and read in the experimental data frames, select the cohort (matched pairs)
```{r message=FALSE, warning=FALSE}
library("readxl")
library("dplyr")
library("xlsx")
library("dplyr")

Base_dir = "D:\\Users\\rpowell\\Documents\\GitHub\\TNBC-PGx"
knitr::opts_knit$set(root.dir = Base_dir)
setwd(Base_dir)


DNA_all <- read.xlsx2("data/TNBC_oncoprint_binary.xlsx",sheetIndex = 1,row.names = 1)
RNA_all <- read.csv("./data/RNAseq_COMBAT.csv",row.names=1)
ssGSEA_all <- read.csv("./data/RNAseq_COMBAT_ssGSEA.csv",row.names=1)

Drug_all <- read.csv("./data/Drug_AUC.csv",row.names=1)
Drug_meta <- read.csv("./data/Drug_metadata.csv",row.names=1)

Meta_all <- read_excel("./data/PDX_metadata.xlsx")
PPI = "./data/STRING.csv"

row.names(Meta_all) = Meta_all$PDX_ID

pairsOnly = T
cellLine_drug = colnames(Drug_all)
drugnames = row.names(Drug_all)

cellLine_RNA = colnames(RNA_all)
#genenames = row.names(RNA_filtered_RWR)

all_samples = intersect(cellLine_drug,cellLine_RNA)

Pairs = Meta_all %>% dplyr::count(PiD)
Pairs = Pairs[which(Pairs$n >= 2),]
Meta_pairsOnly = Meta_all[which(Meta_all$PiD %in% Pairs$PiD),]
row.names(Meta_pairsOnly) = Meta_pairsOnly$PDX_ID

if (pairsOnly){
  train_samples = intersect(all_samples,Meta_pairsOnly$PDX_ID)
} else{train_samples = all_samples}
train_samples

tmp = setdiff(colnames(RNA_all),train_samples)
test_samples = intersect(colnames(Drug_all),tmp)
unknown_samples = setdiff(tmp,colnames(Drug_all))

with_held_RNA = RNA_all[,test_samples]
with_held_Drug = Drug_all[,test_samples]
with_held_common = intersect(colnames(with_held_RNA),colnames(with_held_Drug))

with_held_RNA = with_held_RNA[,with_held_common]
with_held_Drug = with_held_Drug[,with_held_common]
rm("tmp","Pairs")
```
This section generates basic statistics and QC visualizations for RNAseq data.
```{r message=FALSE, warning=FALSE}
library("matrixStats")
RNA_means = rowMeans(as.matrix(RNA_all))
RNA_stdDev = rowSds(as.matrix(RNA_all))
plot(RNA_means,RNA_stdDev)

RNA_filter = which(RNA_means >= .5)
RNA_filtered = RNA_all[RNA_filter,]
RNA_filtered_z = t(scale(t(RNA_filtered), center = T, scale = T))

library("RColorBrewer")

nsamples <- ncol(RNA_filtered)
col <- brewer.pal(nsamples, "Paired")
par(mfrow=c(1,2))
plot(density(as.matrix(RNA_filtered[,1])), col=col[1], lwd=2, ylim=c(0,1), las=2, main="", xlab="")
title(main="Filtered data", xlab="RNAseq")
#abline(v=lcpm.cutoff, lty=3)
for (i in 2:nsamples){
  den <- density(as.matrix(RNA_filtered[,i]))
  lines(den$x, den$y, col=col[i], lwd=2)
}

boxplot(RNA_filtered[,], las=2, col=col, main="")
title(main="Filtered data",ylab="RNAseq")
```
Longitudinal analysis of Vanderbilt sub types.
```{r echo=FALSE}
library("tidyr")
library("dplyr")
library("networkD3")

tmp = as.data.frame(Meta_pairsOnly)[,c("PiD","TNBCtype","Time")]
tmp = reshape(tmp, direction = "wide", idvar = "PiD", timevar = "Time")
colnames(tmp) = c('ID','Pre','Mid','Pst')
tmp = tmp[,c('Pre','Mid','Pst')]

links <-
  tmp %>% 
  mutate(row = row_number()) %>%  # add a row id
  pivot_longer(-row, names_to = "col", values_to = "source") %>%  # gather all columns
  mutate(col = match(col, names(tmp))) %>%  # convert col names to col ids
  mutate(source = paste0(source, '_', col)) %>%  # add col id to node names
  group_by(row) %>%
  mutate(target = lead(source, order_by = col)) %>%  # get target from following node in row
  ungroup() %>% 
  filter(!is.na(target)) %>%  # remove links from last column in original data
  group_by(source, target) %>% 
  summarise(value = n(), .groups = "drop")  # aggregate and count similar links

# create nodes data frame from unque nodes found in links data frame
nodes <- data.frame(id = unique(c(links$source, links$target)),
                    stringsAsFactors = FALSE)

# remove column id from node names
nodes$name <- sub('_[0-9]*$', '', nodes$id)
nodes[nodes=="NA"] = NA

# create node ids in links data to the 0-based index of the nodes in the nodes data frame
links$source_id <- match(links$source, nodes$id) - 1
links$target_id <- match(links$target, nodes$id) - 1

#links = links[intersect(which(!str_detect(links$source,"NA_")),which(!str_detect(links$target,"NA_"))),]

sn = sankeyNetwork(Links = links,
                         Nodes = nodes,
                         Source = 'source_id',
                         Target = 'target_id',
                         Value = 'value',
                         NodeID = 'name',
                         fontSize = 12,
                         iterations = 5)
sn
```

Next, we looked at differential drug susceptibilities identified by the high throughput chemical screening assays. 
```{r}
library(matrixStats)
drug_range = rowRanges(as.matrix(Drug_all),useNames = TRUE)
drug_range[drug_range < 0] <- 0
drug_range = drug_range[,2]-drug_range[,1]

drug_filter = which(drug_range >= 0.5)
```

From the differential gene expression analysis comparing matched pairs, we observed enrichment of cell cycle related genes. This observation is consistent with the growth analysis performed on the ex vivo cultures from the high throughput screens, which show increases in growth rates when comparing the post-NACT to the pre-NACT PDX models. It is also well established that alterations in growth rate and confound PGx associations. Accordingly, in this section we performed a correlative analysis of the growth data to the AUC activities.
```{r}
ord = colnames(Drug_all)
gr = Meta_all[ord,]
test = cor(gr$Log2_FC_3day,t(Drug_all))
r2 = test^2

#get list of correlated and variable drug names
growth_corr = names(r2[,which(r2 > 0.25)])
variable_drugs = names(drug_filter)
x = list(Growth_correlated_drugs = growth_corr,
         Variable_drug_response = variable_drugs)

library(ggvenn)
ggvenn(x,
       fill_color = c("#0073C2FF","#CD534CFF"),
       stroke_size = 0.5, set_name_size = 4)
```
This is the list of variable drugs that are potentially confounded by the rate of growth
```{r}
intersect(growth_corr,variable_drugs)
```
And this is the list of drugs that are not correlated with growth.
```{r}
drug_filter = variable_drugs[!(variable_drugs %in% growth_corr)]
drug_filter[order(drug_filter)]
```
Spot check correlation
```{r}
library("plotly")
df = as.data.frame(cbind(growth=gr$Log2_FC_3day,
                         response=t(Drug_all["PEVONEDISTAT_Broad",])))
colnames(df) = c("growth","response")
ggplot(df,aes(x=growth, y=response))+geom_point()+geom_smooth(method=lm)
```


```{r}
library("matrixStats")
library("ggplot2")

MEAN_AUC = rowMeans(Drug_all)

hc = hclust(dist(Drug_all))
Cluster_ID = cutree(hc,k=4)

df = as.data.frame(cbind(Cluster_ID,MEAN_AUC))
df$Cluster_ID = as.factor(df$Cluster_ID)

ggplot(df, aes(x=Cluster_ID, y=MEAN_AUC, fill=Cluster_ID)) + geom_violin(trim=FALSE) + geom_boxplot(width=0.1)
```

```{r}
library("sunburstR")
library("d3r")

dat = Drug_meta[row.names(df)[which(df$Cluster_ID %in% c(3,4))], ] %>% select("Class","Target")
dat[,"Drug_name"] = row.names(dat)
dat = dat[which(dat$Class != ""),]

dat = data.frame(lapply(dat, function(v) {
  return(toupper(v))
}))

df = dat %>% dplyr::count(Class,Target,Drug_name)
colnames(df)[ncol(df)] <- "size"

tree <- d3_nest(df, value_cols = "size")
sb1 <- sunburst(tree, width="100%", height=400,count = T, sortFunction = htmlwidgets::JS(
      "
    function(a,b) {
      // sort by count descending
      //   unlike the other example using data.name, value is at the top level of the object
      return b.value - a.value
    }
    "
)

)
htmltools::save_html(sb1, file = "./Clus4_panActive.html")
sb1
```


```{r}
library("dendextend")
library("ComplexHeatmap")

CCL_corr = cor(as.matrix(t(scale(t(Drug_all[drug_filter,train_samples]),center=TRUE, scale=TRUE))),method = c("pearson"))
df_anno = Meta_all[train_samples,]%>%select("Time","PiD")

K = 7
hc = hclust(dist(CCL_corr))
clus = cutree(hc,k=K)

df_anno_clust = cbind(df_anno,clus)
chisq.test(df_anno_clust$PiD, df_anno_clust$clus, correct=FALSE)
table(df_anno_clust$PiD, df_anno_clust$clus)

row_dend = as.dendrogram(hc)
row_dend = color_branches(row_dend, k = K)

HM = Heatmap(CCL_corr,
             name = "Pearson",
             top_annotation = anno_gen(df_anno),
             show_row_names = FALSE)
HM
```

Differential drug susceptibility analysis
```{r warning=FALSE}
library(edgeR)
metadata = Meta_all[train_samples,]
metadata[] <- lapply(metadata, factor)

design = model.matrix(~0+Time+PiD,metadata)
Rx_fit = lmFit(Drug_all[drug_filter,train_samples],design)
Rx_fit_eBayesfit = eBayes(Rx_fit,trend = TRUE,robust = TRUE)
Rx_top = topTable(Rx_fit_eBayesfit,coef = "TimePRE" ,adjust.method="fdr", number=Inf)

Mid_v_Pre = c()
Pst_v_Pre = c()
for (i in 1:nrow(Drug_all)){
  tmp1 = t(Drug_all[i,train_samples])
  colnames(tmp1) = "Active"
  
  tmp2 = Meta_all[train_samples,]
  tmp3 = cbind(tmp1,tmp2)
  tmp4 = tmp3[,c("Active","Time","PiD")]
  
  tmp5 = reshape(tmp4, direction = "wide", idvar = "PiD", timevar = "Time")
  tmp6 = tmp5$Active.MID-tmp5$Active.PRE
  tmp7 = tmp5$Active.POST-tmp5$Active.PRE
  names(tmp6) = tmp5$PiD
  names(tmp7) = tmp5$PiD
  
  Mid_v_Pre = rbind(Mid_v_Pre,tmp6)
  Pst_v_Pre = rbind(Pst_v_Pre,tmp7)
}

rm(list = c("tmp1","tmp2","tmp3","tmp4","tmp5","tmp6","tmp7"))

rownames(Mid_v_Pre) = row.names(Drug_all)
Mid_v_Pre = Mid_v_Pre[ , colSums(is.na(Mid_v_Pre)) == 0]
Mid_v_Pre = Mid_v_Pre[rownames(Rx_top),]

rownames(Pst_v_Pre) = row.names(Drug_all)
Pst_v_Pre = Pst_v_Pre[ , colSums(is.na(Pst_v_Pre)) == 0]
Pst_v_Pre = Pst_v_Pre[rownames(Rx_top),]

Rx_top = cbind(Rx_top,shift=rowMeans(cbind(Mid_v_Pre,Pst_v_Pre)))
Rx_top

write.csv(Rx_top,"./Reports/Drug_LmPiD_AUC_notPDXpre.csv")

thresh_Pval = 0.05
thresh_FC = 0.05

tmp = Rx_top[intersect(which(Rx_top$P.Value <= thresh_Pval),which(abs(Rx_top$shift) >= thresh_FC)),]
up = as.integer(length(which(tmp$shift > 0)))
down =  as.integer(length(which(tmp$shift <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = Rx_top, x = ~shift, y = ~-1*log10(P.Value))%>%
  layout(xaxis = list(title = "Delta"),yaxis = list(title = "-log(p-value)"),
         shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(Rx_top),marker = list(color='black'))

plotly::export(p=fig,file = "Reports/Drug_LmPiD_AUC_notPDXpre.png")
fig
```

```{r fig.height=20, fig.width=32}
library("ComplexHeatmap")
library("dplyr")

group_df = as.data.frame(Drug_meta[row.names(Rx_top),] %>% dplyr::select("Class"),use_raster = TRUE, raster_by_magick = TRUE)
anno_df = as.data.frame(Drug_meta[row.names(Rx_top),] %>% select("Target"),use_raster = TRUE, raster_by_magick = TRUE)
anno_df = cbind(X1_v_X0=rowMeans(Mid_v_Pre),XS_v_X0=rowMeans(Pst_v_Pre),anno_df)
anno_txt = anno_text(anno_df$Target, which = "row")

library(circlize)
col_fun = colorRamp2(c(-0.33,0,0.33), c("blue","white","red"))
#col_fun(seq(-10, 10))

rAnno = rowAnnotation(df = anno_df,
                      col=list(X1_v_X0=col_fun,XS_v_X0=col_fun), 
                      width = ncol(anno_df)*unit(10, "mm")
                      )

hm1 = Heatmap(Mid_v_Pre[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0.05)),],
              width = ncol(Mid_v_Pre)*unit(20, "mm"),
              name = "dAUC",
              clustering_method_rows = "average",
              column_title = "MID-PRE",
              column_title_gp=gpar(fontsize=20,fontface = "bold"),
              column_names_gp = gpar(fontsize = 24),
              row_split = group_df$Class[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0.05))],
              gap = unit(2,"mm"),
              row_title_rot = 0,
              row_title_gp=gpar(fontsize=24,fontface = "bold"),
              heatmap_legend_param = list(
                legend_direction = "horizontal",
                legend_width = unit(6, "cm"),
                title_position = "topcenter",
                title_gp = gpar(fontsize = 24),
                labels_gp = gpar(fontsize = 24)),
              col = col_fun,
              use_raster = TRUE, raster_by_magick = TRUE)

hm2 = Heatmap(Pst_v_Pre[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0.05)),],
              width = ncol(Pst_v_Pre)*unit(20, "mm"),
              name = "dAUC",
              column_title = "POST-PRE",
              column_title_gp=gpar(fontsize=20,fontface = "bold"),
              column_names_gp = gpar(fontsize = 24),
              heatmap_legend_param = list(
                legend_direction = "horizontal",
                legend_width = unit(6, "cm"),
                title_position = "topcenter",
                title_gp = gpar(fontsize = 24),
                labels_gp = gpar(fontsize = 24)),
              row_labels = anno_df$Target[intersect(which(Rx_top$adj.P.Val <= 0.05),which(abs(Rx_top$shift) >= 0.05))],
              row_names_gp = gpar(fontsize = 24),
              col = col_fun,
              use_raster = TRUE, raster_by_magick = TRUE)

ht_list = hm1+hm2+anno_txt

draw(ht_list,
     row_title = "dAUC",
     row_title_gp = gpar(col = "black"),
     heatmap_legend_side = "bottom",
     annotation_legend_side = "right",
     legend_labels_gp = gpar(36))
```


Next, we evaluated the similarity of the matched pairs to determine whether individual patients or time points tended to associate with one another. From this analysis we found that individual patients tended to co-cluster over time points. Thus recapitulating a strong degree of patient heterogeneity.
```{r echo=FALSE}
library("dendextend")
library("ComplexHeatmap")

CCL_corr = cor(as.matrix(RNA_filtered_z[,train_samples]),method = c("pearson"))
df_anno = Meta_all[train_samples,]%>%select("Time","PiD")

K = 9
hc = hclust(dist(CCL_corr))
clus = cutree(hc,k=K)

df_anno_clust = cbind(df_anno,clus)
chisq.test(df_anno_clust$PiD, df_anno_clust$clus, correct=FALSE)
table(df_anno_clust$PiD, df_anno_clust$clus)

row_dend = as.dendrogram(hc)
row_dend = color_branches(row_dend, k = K)

HM = Heatmap(CCL_corr,
             name = "Pearson",
             cluster_rows = row_dend,
             cluster_columns = row_dend,
             top_annotation = anno_gen(df_anno),
             show_row_names = FALSE)
HM
```
We then perform differential gene expression analysis that accounted for the individual patient matched pairs. 
```{r warning=FALSE}
library("edgeR")
library("plotly")

metadata = Meta_all[train_samples,]
metadata[] <- lapply(metadata, factor)

design = model.matrix(~0+Time+PiD,metadata)
DEG_fit = lmFit(RNA_filtered_z[,train_samples],design)
cm = makeContrasts(Mid_v_Pre=TimeMID-TimePRE,levels=design)
DEG_fit_cm = contrasts.fit(DEG_fit,cm)
DEG_fit_cm_eBayesfit = eBayes(DEG_fit_cm,trend = FALSE,robust = FALSE)
DEG_top = topTable(DEG_fit_cm_eBayesfit,adjust.method="fdr", number=Inf)

# DEG_fit_cm_eBayesfit = eBayes(DEG_fit,trend = FALSE,robust = FALSE)
# DEG_top = topTable(DEG_fit_cm_eBayesfit, coef = "TimePRE",adjust.method="fdr", number=Inf)

DEG_top
write.csv(DEG_top,"./Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre.csv")

thresh_Pval = 0.05
thresh_FC = 0

tmp = DEG_top[intersect(which(DEG_top$P.Value <= thresh_Pval),which(abs(DEG_top$logFC) >= thresh_FC)),]
up = as.integer(length(which(tmp$logFC > 0)))
down =  as.integer(length(which(tmp$logFC <= 0)))
print(paste("Total DEG: ", up+down))
print(paste("Sigificantly up-reglated genes: ", up))
print(paste("Sigificantly down-reglated genes: ", down))
rm("tmp","up","down")


#Volcano plot of DEG
fig <- plot_ly(data = DEG_top, x = ~logFC, y = ~-1*log10(P.Value),color = ~logFC, colors="RdBu")%>%
  layout(xaxis = list(title = "DeltaZ", tickfont = list(size = 40),titlefont = list(size = 60)),
         yaxis = list(title = "-log(p-value)", tickfont = list(size = 40),titlefont = list(size = 60)),
         shapes=list(vline(-1*thresh_FC),vline(thresh_FC),hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(DEG_top))

plotly::export(p=fig,file = "Reports/DEG_LmPID_RNAz_PDXmid_v_PDXpre.png")
  
fig
```

Gene-pathway enrichment analysis using the DGE list from the longitudinal matched pairs.
```{r message=TRUE, warning=FALSE}
library("pathfindR")
library("visNetwork")

pfR_input = DEG_top[,c("logFC","P.Value"),]
Gene_symbol = row.names(pfR_input)
pfR_input = cbind(Gene_symbol,pfR_input)

name = "Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre_pathfindR"

#Delete old
unlink(name, recursive = TRUE)

pfr_tmp <- run_pathfindR(pfR_input,
                               iterations = 10,
                               p_val_threshold = 0.05,
                               search_method = "GR",
                               #gene_sets = "Reactome",
                               #pin_name_path = "STRING",
                               plot_enrichment_chart = FALSE,
                               visualize_enriched_terms = FALSE,
                               output_dir = name,
                               silent_option = TRUE)

pfr_tmp <- pfr_tmp[order(pfr_tmp$lowest_p, decreasing = FALSE), ]
assign(paste("pfr_",name,sep=""),pfr_tmp)

pathfindR_visNetwork(pfr_tmp,paste(name,"visNetwork.html",sep="/"))
hsa_preview(pfr_tmp,pfR_input,outdir = paste(Base_dir,name,"hsa_preview",sep="/"))

rm("pfr_tmp")
```

```{r fig.height=14, fig.width=12}
library(stringr)
library(tidyr)

colfunc <- colorRampPalette(c("black", "white"))

tmp = `pfr_Reports/DEG_LmPiD_RNAz_PDXmid_v_PDXpre_pathfindR`[1:10,]
#Volcano plot of DEG
plot_DGE <- plot_ly(data = tmp,x = ~Fold_Enrichment, y = ~Term_Description, color = ~-log10(lowest_p), size = ~occurrence, type = "scatter", marker=list(line=list(color="black")))%>%layout(xaxis = list(title = "Enrichment plot"), yaxis = list(title = "",tickfont = list(size = 20))) #, categoryorder = "total ascending")
plot_DGE

df = data.frame()
for (i in 1:nrow(tmp)){
  df[i,"Term"] = tmp$Term_Description [i]
  df[i,"positive_corr"] = length(unlist(str_split(tmp$Up_regulated[i],",")))
  df[i,"negative_corr"] = length(unlist(str_split(tmp$Down_regulated[i],",")))
}

df$negative_corr = -1*df$negative_corr

df.long <- df[1:10,] %>% pivot_longer(-Term, names_to = "variable", values_to = "value")

df.long %>% plot_ly(x= ~value, y=~Term, color=~variable) %>% 
  layout(bargap = 0.1, barmode = 'overlay',
         xaxis = list(title = "N Genes", tickfont = list(size = 20)),
         yaxis = list(title = "",tickfont = list(size = 20)),
         legend = list(title = "Interaction",tickfont = list(size = 20))) #%>% add_bars(orientation = 'h', hoverinfo = 'text', text = ~Term)
```

```{r}
Y_all = t(Drug_all[drug_filter,train_samples])
drugnames = colnames(Y_all)

#X_all = t(RNA_filtered[,common])
X_all = t(RNA_filtered_z[,train_samples])
```

```{r}
library(fgsea)
library(corto)
pathways = gmtPathways("R:\\TAMU-IBT_Powell\\Code\\R_scripts\\ssGSEA2.0-master\\db\\msigdb\\canonical.gmt")
```

```{r}
sel = c("PIM137","PIM172",
        "PIM025","PIM038",
        "PIM190","PIM231",
        "PIM254","PIM311")
PEV_ = c("RESP","NON",
          "NON","NON",
          "NON","NON",
          "RESP","RESP")

library(dplyr)
library(tidyr)
library(rstatix)
genes = pathways[["REACTOME_NEDDYLATION"]]
genes = intersect(genes,row.names(RNA_filtered_z))

df = as.data.frame(t(rbind(Gx =PEV_,RNA_filtered_z[,sel])))
df.long <- df %>% pivot_longer(-Gx, names_to = "genes", values_to = "value")

df.long$Gx = as.factor(df.long$Gx)
df.long$value = as.numeric(df.long$value)  
  
stat.test <- df.long %>%
  group_by(genes) %>%
  t_test(value ~ Gx,detailed =TRUE) %>%
  adjust_pvalue(method = "BH") %>%
  add_significance()

stat.test[,c("genes","estimate","p","p.adj")]
write.csv(stat.test[,c("genes","estimate","p","p.adj")],file="Reports/Ttest_Resp2Non_RNAz.csv")
```

```{r}
library("pathfindR")
library("stringr")
library("dplyr")

InVivo_idx = which(!is.na(Meta_all$InVivo_PEV))
InVivo_list = Meta_all[InVivo_idx,"InVivo_PEV"]
InVivo_models = Meta_all[InVivo_idx,"PDX_ID"]

tmp = X_all[unlist(as.matrix(InVivo_models)),]
tmp = cbind(InVivo_list,tmp)
tmp = apply(tmp[, -1], 2, function(x) {cor.test(tmp[, 1], x, method = "pearson", use = "pairwise")})

p.vals <- as.matrix(unlist(sapply(tmp, "[[", "p.value")))
r.vals <- as.matrix(unlist(sapply(tmp, "[[", "estimate")))

tmp = as.data.frame(cbind(colnames(X_all),
                          r_val = r.vals,
                          p_val = p.vals))
colnames(tmp) = c("gene_symbol","pearson","P_value")
rownames(tmp) <- NULL

tmp$pearson = as.numeric(tmp$pearson)
tmp$P_value = as.numeric(tmp$P_value)

#Volcano plot of DEG
fig <- plot_ly(data = tmp, x = ~pearson, y = ~-1*log10(P_value),color = ~pearson, colors="RdBu")%>%
  layout(xaxis = list(title = "Correlation", tickfont = list(size = 30),titlefont = list(size = 35)),
         yaxis = list(title = "-log(p-value)", tickfont = list(size = 30),titlefont = list(size = 35)),
         shapes=list(hline(-log10(thresh_Pval)))) %>%
  add_trace(text=row.names(tmp))

plotly::export(p=fig,file = "Reports/Cor_pearson_TC_RNAz.png")
fig

# write.csv(tmp, file="Reports/Cor_pearson_TC_RNAz.csv")
# 
# name = "Reports/Cor_pearson_TC_RNAz_pathfindR_2"
# 
# #Delete old
# unlink(name, recursive = TRUE)
# 
# pfr_tmp <- run_pathfindR(tmp,
#                                iterations = 10,
#                                p_val_threshold = 0.05,
#                                search_method = "GR",
#                                #gene_sets = "Reactome",
#                                #pin_name_path = "STRING",
#                                plot_enrichment_chart = FALSE,
#                                visualize_enriched_terms = FALSE,
#                                output_dir = name,
#                                silent_option = FALSE)
# 
# pfr_tmp <- pfr_tmp[order(pfr_tmp$lowest_p, decreasing = FALSE), ]
# assign(paste("pfr_",name,sep=""),pfr_tmp)
# 
# rownames(tmp) <- tmp$gene_symbol
# colnames(tmp) = c("gene_symbol","LogFC","P_value")
# 
# pathfindR_visNetwork(pfr_tmp,paste(name,"visNetwork.html",sep="/"))
# hsa_preview(pfr_tmp,tmp[which(tmp$P_value<=0.05),],outdir = paste(Base_dir,name,"hsa_preview",sep="/"))

#rm("pfr_tmp")
```


```{r fig.height=4, fig.width=8}
library(plotly)
tmp = `pfr_Reports/Cor_pearson_TC_RNAz_pathfindR`[order(pfr_tmp$lowest_p,decreasing = F),]
tmp = tmp[1:10,]
#Volcano plot of DEG
plot_DGE <- plot_ly(data = tmp,x = ~Fold_Enrichment, y = ~Term_Description, color = ~-log10(lowest_p), colors="Blues", size = ~occurrence, type = "scatter", marker=list(line=list(color="black")))%>%
  layout(xaxis = list(title = "Enrichment plot", tickfont = list(size = 30),titlefont = list(size = 35)),
         yaxis = list(title = "", categoryorder = "total ascending", tickfont = list(size = 30),titlefont = list(size = 35)),
         width = 1600, height = 800,
         legend = list(title=list(text='p_value'), font = list(size = 30)),
         legend_title = list(font = list(size = 30))
         )
plot_DGE
plotly::export(p=plot_DGE,file = "Reports/Cor_pearson_TC_RNAz_pathfindR_enrichment.png")

df = data.frame()
for (i in 1:nrow(tmp)){
  df[i,"Term"] = tmp$Term_Description [i]
  df[i,"up_reglated"] = length(unlist(str_split(tmp$Up_regulated[i],",")))
  df[i,"down_reglated"] = length(unlist(str_split(tmp$Down_regulated[i],",")))
}

df$down_reglated = -1*df$down_reglated

df.long <- df[1:10,] %>% pivot_longer(-Term, names_to = "variable", values_to = "value")

fig = df.long %>% plot_ly(x= ~value, y=~Term, color=~variable) %>% 
  layout(bargap = 0.1, barmode = 'overlay',
         xaxis = list(title = "N Genes", tickfont = list(size = 20),titlefont = list(size = 24)),
         yaxis = list(title = "", tickfont = list(size = 20),titlefont = list(size = 24))
         )
fig
plotly::export(p=fig,file = "Reports/Cor_pearson_TC_RNAz_pathfindR_gene_counts.png")
```


```{r}
save.image(file='TNBC-PGx_analysis.RData')
sessionInfo()
```
